File: core/src/main/java/io/github/frikit/krandom/generator/base/LongGenerator.java
/*
 * Copyright (c) 2026 krandom contributors
 *
 * Licensed under the MIT License. See LICENSE in the project root for license information.
 */
package io.github.frikit.krandom.generator.base;

/**
 * Generates random {@link Long} values.
 *
 * <p>Default range: [{@code Long.MIN_VALUE}, {@code Long.MAX_VALUE}) — the full 64-bit range
 * minus the maximum value itself (exclusive upper bound).
 *
 * <pre>{@code
 *   long any       = new LongGenerator().generate();
 *   long timestamp = new LongGenerator(0L, System.currentTimeMillis()).generate();
 * }</pre>
 */
public final class LongGenerator extends AbstractBoundedGenerator<Long> {

    public LongGenerator() {
        super(Long.MIN_VALUE, Long.MAX_VALUE, null);
    }

    public LongGenerator(long min, long max) {
        super(min, max, null);
    }

    public LongGenerator(long min, long max, long seed) {
        super(min, max, seed);
    }

    /**
     * Generate a long in the half-open range [{@code min}, {@code max}).
     *
     * @throws IllegalArgumentException if {@code min >= max}
     */
    @Override
    public Long generate(Long min, Long max) {
        validate(min, max);
                return random.nextLong(min, max);
    }
}

File: core/src/main/kotlin/org/github/krandom/common/generators/LongGenerator.kt
/*
 * Copyright (c) 2026 krandom contributors
 *
 * Licensed under the MIT License. See LICENSE in the project root for license information.
 */
package org.github.krandom.common.generators

import org.github.krandom.common.RangedTypeGenerator
import java.security.SecureRandom
import java.util.Random

/**
 * Generates random [Long] values.
 *
 * - [generate()] generates within the configured [start, end) range (defaults: full Long range).
 * - [generate(start, end)] uses a half-open range [start, end): start is inclusive, end is exclusive.
 */
class LongGenerator(
    private val start: Long = Long.MIN_VALUE,
    private val end: Long = Long.MAX_VALUE,
    seed: Long? = null
) : RangedTypeGenerator<Long> {

    private val random: Random = if (seed != null) Random(seed) else SecureRandom()

    override fun generate(): Long = generate(start, end)

    override fun generate(start: Long, end: Long): Long {
        require(start != end) { "start ($start) and end ($end) must differ" }
        val lo = minOf(start, end)
        val hi = maxOf(start, end)
        return random.longs(1, lo, hi).findFirst().orElse(lo)
    }
}
